### PR TITLE
Accept regexps in rowsum() to make row specification easier

### PR DESCRIPTION
**Pull request by redstreet**

### CODE REVIEW

**Comment by redstreet** [comment]
Apologies for the PR noise.

File: 2022/ca540sca.py [hunk 1]
```diff
@@ -16,9 +16,8 @@ def __init__(f, inputs, f1040, f1040sa):
                                 '1A_B6b', '1A_B7', '1B_B1', '1B_B3', '1B_B4',
                                 '1B_B5', '1B_B6', '1B_B7', '1B_B9a', '1B_B9b1',
                                 '1B_B9b2', '1B_B9b3'])
-        f['1B_C10'] = f.rowsum(['1A_C1z', '1A_C2b', '1A_C3b', '1A_C4b', '1A_C5b',
-                                '1A_C7', '1B_C2a', '1B_C3', '1B_C4', '1B_C5',
-                                '1B_C6', '1B_C9a'])
+        f['1B_C10'] = f.rowsum(['1A_C7', '1B_C2a', '1B_C9a'],
+                               patterns=['1A_C1[a-z]', '1A_C[2345]b', '1B_C[3-6]'])
```
Comment (davidcmoore) [inline 2022/ca540sca.py:20]: This isn't quite correct since 1z is a subtotal of 1a through 1i.

File: 2022/form.py [hunk 2]
```diff
@@ -74,7 +74,7 @@ def mustFile(self):
         """
         return self.must_file
 
-    def rowsum(self, rows):
+    def rowsum(self, rows, patterns=[]):
```
Comment (davidcmoore) [inline 2022/form.py:77]: I like this idea but I wonder if we could just make `rows` support regexs always and drop the need for an extra argument. That should be fine as long as none of the existing uses could be misinterpreted as regexes.

File: 2022/form.py [hunk 3]
```diff
@@ -85,6 +85,11 @@ def rowsum(self, rows):
             if r in self:
                 isNone = False
                 val += self[r]
+
+        for p in patterns:
+            pat = re.compile(p)
+            val += sum(v for k, v in self.data.items() if pat.match(k))
```
Comment (davidcmoore) [inline 2022/form.py:91]: We need to make sure to set isNone=False if there are any matches.

**Review by davidcmoore — CHANGES_REQUESTED** [review]

File: 2022/form.py (hunk 3 above)
Comment (davidcmoore) [reply 2022/form.py:91]: Also should use `fullmatch()` because I think substring matches are not intended here.

File: 2022/ca540sca.py (hunk 1 above)
Comment (redstreet) [reply 2022/ca540sca.py:20]: Good catch, fixed to `a-i`.

File: 2022/form.py (hunk 2 above)
Comment (redstreet) [reply 2022/form.py]: That's much cleaner. Done..

> That should be fine as long as none of the existing uses could be misinterpreted as regexes.

To test this and others:
- I looked through a `git grep rowsum` for one year and didn't notice anything that could be misinterpreted as a regex
- I ran the last 3 years of my own taxes, and nothing changed

**Performance:**
Ran this 5 times each for my 2022 taxes, and took the average (variance was low):
```
    start = time.perf_counter()
    f = F1040(inputs)
    ca = CA540(inputs, f)
    end = time.perf_counter()
```

```
0.00046s original code
0.00667s matching each pattern individually
0.00517s using | to combine regexes
```

So there is a 11x performance decrease. But given the absolute numbers are so small, and given I can't imagine someone running a few thousand tax returns at once, I'd think this makes no difference practically. WDYT?

**Comment by redstreet** [comment]
Quick ping whenever you get a chance :).

**Review by davidcmoore — APPROVED** [review]
Sorry for the delay, my bad! Yes, this looks good now.

**Pull request merged**